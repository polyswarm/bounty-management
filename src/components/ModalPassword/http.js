class HttpAccount {
  constructor(url) {
    this.url = url;
  }

  unlockWallet(address, password) {
    const url = this.url;
    return fetch(url+'/accounts/'+address+'/unlock', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({'password': password})
    })
      .then(response => response.ok)
      .catch(() => false);
  }

  createWallet(password) {
    const url = this.url;
    return fetch(url + '/accounts', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({'password': password})
    })
      .then(response => response.ok)
      .catch(() => false);
  }

  getEth(address) {
    const url = this.url;
    return fetch(url+'/accounts/'+address+'/balance/eth')
      .then(response => {
        if (response.ok) {
          return response;
        }
        throw Error('Failed to get balance');
      })
      .then(response => response.json())
      .then(json => json.result+'')
      .catch(() => 0);
  }

  getNct(address) {
    const url = this.url;
    return fetch(url+'/accounts/'+address+'/balance/nct')
      .then(response => {
        if (response.ok) {
          return response;
        }
        throw Error('Failed to get balance');
      })
      .then(response => response.json())
      .then(json => json.result+'')
      .catch(() => 0);
  }
}
export default HttpAccount;
