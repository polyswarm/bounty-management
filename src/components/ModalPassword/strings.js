const strings = {
  password: 'Wallet Password',
  store: 'Keep this wallet unlocked',
  error: 'Unabled to login. Check your password.',
  cancel: 'Cancel',
  unlock: 'Unlock',
  header: 'Unlock your wallet.',
  createHeader: 'Create a wallet',
  create: 'Create',
  address: 'Wallet Address'
};

export default strings;
