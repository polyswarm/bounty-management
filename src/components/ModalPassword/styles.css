.ModalBackground {
  background-color: #000000b0;
  display: flex;
  flex-direction: column;
  height: 100vh;
  justify-content: center;
  left: 0;
  position: fixed;
  top: 0;
  width: 100vw;
  z-index: 1;
}

.ModalContent {
  background-color: white;
  box-shadow: 0 0 10px black;
  box-sizing: border-box;
  color: black;
  display: flex;
  flex-direction: column;
  height: 350px;
  justify-content: flex-start;
  left: 50%;
  margin: auto;
  position: absolute;
  top: 50%;
  transform: translate(-50%, -50%);
  width: 300px;
  z-index: 2;
}

.ModalContent > form {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  padding: 20px;
}

.ModalContentHeader {
  background-color: #8500ff;
  box-sizing: border-box;
  color: white;
  display: flex;
  flex-direction: column;
  font-size: 24px;
  font-weight: 900;
  height: 60px;
  justify-content: center;
  padding: 0 20px;
  text-align: center;
  width: 100%;
}

.Modal-Button-Bar {
  bottom: 0;
  position: absolute;
  width: 100%;
}

.Modal-Button-Bar > button{
  float: right;
  margin-bottom: 5px;
  width: 110px;
}

.ModalError {
  color: #ff1a00;
}

select {
  background-color: white;
  font-size: 16px;
  margin-bottom: 10px;
  padding: 4px 12px;
}

select:focus {
  outline-color: #8500ff;
}

input[type=password] {
  box-sizing: border-box;
  font-size: 16px;
  margin-bottom: 10px;
  padding: 4px 12px;
}

input[type=password]:focus {
  outline-color: #8500ff;
}
