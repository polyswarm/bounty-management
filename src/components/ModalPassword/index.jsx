// Vendor imports
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import BigNumber from 'bignumber.js';
import Uuid from 'uuid/v4';
// Bounty imports
import Button from '../Button';
// Component imports
import HttpAccount from './http';
import strings from './strings';

class ModalPassword extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      unlocking: false,
      error: false,
      password: '',
      address: 0,
      eth: 0,
      nct: 0
    };

    this.onWalletChangeHandler = this.onWalletChangeHandler.bind(this);
    this.onChangeStore = this.onChangeStore.bind(this);
    this.onChangePassword = this.onChangePassword.bind(this);
    this.onChangeAddress = this.onChangeAddress.bind(this);
    this.onCloseClick = this.onCloseClick.bind(this);
    this.onKeyPress = this.onKeyPress.bind(this);
    this.onUnlockClick = this.onUnlockClick.bind(this);
    this.unlockWallet = this.unlockWallet.bind(this);
    this.createWallet = this.createWallet.bind(this);
    this.open = this.open.bind(this);
    this.close = this.close.bind(this);
    this.updateBalance = this.updateBalance.bind(this);
    this.addAccountRequest = this.addAccountRequest.bind(this);
    this.removeAccountRequest = this.removeAccountRequest.bind(this);
  }

  componentWillMount() {
    const { props: { walletList } } = this;
    const { state: { address } } = this;
    if (walletList && walletList.length > 0) {
      this.updateBalance(walletList[address]);
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { props: { walletList } } = this;
    const { state: { address } } = this;
    const { address: prevAddr } = prevState;
    const { walletList: prevWallets } = prevProps;

    if (walletList
      && walletList.length > 0 
      && (prevAddr !== address 
      || JSON.stringify(walletList) !== JSON.stringify(prevWallets))) {
      this.updateBalance(walletList[address]);
    }
  }

  render() {
    const { props: { walletList } } = this;
    const {
      state: { open, unlocking, error, password, address, nct, eth }
    } = this;
    return (
      <div className='ModalPassword'>
        {open && (
          <React.Fragment>
            <div className='ModalBackground' onClick={this.onCloseClick} />
            <div className='ModalContent'>
              <header className='ModalContentHeader'>
                {walletList.length > 0 ? strings.header : strings.createHeader}
              </header>
              <form>
                {walletList.length > 0 && (
                  <React.Fragment>
                    <label htmlFor='address'>{strings.address}</label>
                    <select
                      id='address'
                      value={walletList[address]}
                      onChange={this.onChangeAddress}>
                      {walletList.map(wallet => {
                        return (
                          <option key={wallet} value={wallet}>
                            {wallet}
                          </option>
                        );
                      })}
                    </select>
                    <label>Balances</label>
                    <div className='Balances'>
                      <p className='Balance'>NCT: {nct}</p>
                      <p className='Balance'>ETH: {eth}</p>
                    </div>
                  </React.Fragment>
                )}
                <label htmlFor='password'>{strings.password}</label>
                <input
                  id='password'
                  type='password'
                  value={password}
                  onKeyPress={this.onKeyPress}
                  onChange={this.onChangePassword}
                />
                <div className='ModalError'>{error && strings.error}</div>
              </form>
              <p className='ModalMessage'>{strings.background}</p>
              <span className='Modal-Button-Bar'>
                <Button flat
                  disabled={unlocking} 
                  onClick={this.onUnlockClick}>
                  {walletList.length > 0 ? strings.unlock : strings.create}
                </Button>
                <Button
                  flat
                  cancel
                  disabled={unlocking}
                  onClick={this.onCloseClick}>
                  {strings.cancel}
                </Button>
              </span>
            </div>
          </React.Fragment>
        )}
      </div>
    );
  }

  onWalletChangeHandler() {
    const { props: { onWalletChange }, state: { store } } = this;
    if (onWalletChange) {
      onWalletChange(store);
    }
  }

  onChangeStore(event) {
    this.setState({ store: event.target.checked });
  }

  onChangePassword(event) {
    this.setState({ password: event.target.value });
  }

  onChangeAddress(event) {
    const { props: { walletList } } = this;
    const value = event.target.value;
    const index = walletList.findIndex(v => v === value);
    this.setState({ address: index });
  }

  onCloseClick() {
    const { state: { unlocking } } = this;
    if (!unlocking) {
      this.close();
    }
  }

  onKeyPress(e) {
    if (e.key == 'Enter') {
      e.preventDefault();
      this.onUnlockClick();
    }
  }

  onUnlockClick() {
    const { state: { address, password } } = this;
    const { props: { walletList } } = this;
    if (walletList && walletList.length > 0) {
      this.unlockWallet(walletList[address], password);
    } else {
      this.createWallet(password);
    }
  }

  unlockWallet(address, password) {
    const { props: { url } } = this;
    this.setState({ unlocking: true, error: false });
    const http = new HttpAccount(url);
    const uuid = Uuid();
    this.addAccountRequest(strings.requestUnlockWallet, uuid);
    return http.unlockWallet(address, password).then(success => {
      this.setState({ unlocking: false, error: !success });
      if (success) {
        this.onWalletChangeHandler();
        this.close();
      }
      this.removeAccountRequest(strings.requestUnlockWallet, uuid);
    });
  }

  createWallet(password) {
    const { props: { url } } = this;
    this.setState({ unlocking: true, error: false });
    const http = new HttpAccount(url);
    const uuid = Uuid();
    this.addAccountRequest(strings.requestCreateWallet, uuid);
    return http.createWallet(password).then(success => {
      this.setState({ unlocking: false, error: !success });
      if (success) {
        this.onWalletChangeHandler();
        this.close();
      }
      this.removeAccountRequest(strings.requestCreateWallet, uuid);
    });
  }

  open() {
    this.setState({ open: true });
  }

  close() {
    this.setState({ 
      open: false,
      password: ''
    });
  }

  addAccountRequest(title, id) {
    const { addRequest } = this.props;
    if (addRequest) {
      addRequest(title, id);
    }
  }

  removeAccountRequest(title, id) {
    const { removeRequest } = this.props;
    if (removeRequest) {
      removeRequest(title, id);
    }
  }

  updateBalance(address) {
    const { props: { url } } = this;
    const http = new HttpAccount(url);
    const uuid = Uuid();
    this.addAccountRequest(strings.requestBalance, uuid);
    
    const e = http.getEth(address).then(balance => {
      return new BigNumber(balance).dividedBy(new BigNumber(1000000000000000000));
    }).then((b) => {
      this.setState({ eth: b.toNumber() });
    });
    const n = http.getNct(address).then(balance => {
      return new BigNumber(balance).dividedBy(new BigNumber(1000000000000000000));
    }).then((b) => {
      this.setState({ nct: b.toNumber() });
    });
    const promises = [e, n];
    return Promise.all(promises).then(() => {
      this.removeAccountRequest(strings.requestBalance, uuid);
    });
  }
}

ModalPassword.proptypes = {
  url: PropTypes.string,
  walletList: PropTypes.array,
  onWalletChange: PropTypes.func,
  addRequest: PropTypes.func,
  removeRequest: PropTypes.func
};
export default ModalPassword;
