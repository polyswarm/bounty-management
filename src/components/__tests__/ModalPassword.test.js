import React from 'react';
import {render, mount} from 'enzyme';
import {renderToJson} from 'enzyme-to-json';
import ModalPassword from '../ModalPassword';
import HttpAccount from '../ModalPassword/http';

const mockUnlockWallet = jest.fn().mockImplementation(() => {
  return new Promise(resolve => {
    resolve(true);
  });
});

const mockCreateWallet = jest.fn().mockImplementation(() => {
  return new Promise(resolve => {
    resolve(true);
  });
});

jest.mock('../ModalPassword/http', () => {
  // Works and lets you check for constructor calls:
  return jest.fn().mockImplementation(() => {
    return {
      unlockWallet: mockUnlockWallet,
      createWallet: mockCreateWallet,
    };
  });
});

beforeEach(() => {
  jest.clearAllMocks();
  HttpAccount.mockClear();
  mockUnlockWallet.mockClear();
  mockCreateWallet.mockClear();

  HttpAccount.mockImplementation(() => {
    return {
      unlockWallet: mockUnlockWallet,
      createWallet: mockCreateWallet,
    };
  });
});

it('renders without crashing', () => {
  const walletList = [];
  const wrapper = render(<ModalPassword walletList={walletList}/>);
  expect(renderToJson(wrapper)).toMatchSnapshot();
});

it('sets state open to true when open is called', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  const instance = wrapper.instance();
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');

  instance.open();

  expect(setState).toHaveBeenCalledTimes(1);
  expect(setState).toHaveBeenCalledWith({open: true});
});

it('sets state open to false when close is called', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  const instance = wrapper.instance();
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');

  instance.close();

  expect(setState).toHaveBeenCalledTimes(1);
  expect(setState).toHaveBeenCalledWith({open: false});
});

it('hides the Modal when closed', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({open: false});

  expect(wrapper.find('.ModalBackground')).toHaveLength(0);
});

it('shows the Modal when open', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({open: true});

  expect(wrapper.find('.ModalBackground')).toHaveLength(1);
});

it('closes the modal on click outside the main content', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({open: true});

  wrapper.find('.ModalBackground').simulate('click');

  expect(wrapper.find('.ModalBackground')).toHaveLength(0);
});

it('hides the select when wallet list is empty', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({open: true});

  expect(wrapper.find('select')).toHaveLength(0);
});

it('shows the select when wallet list is not empty', () => {
  const walletList = ['asdf'];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({open: true});

  expect(wrapper.find('select')).toHaveLength(1);
});

it('calls onWalletChange with true when store is true', () => {
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword walletList={walletList}
      onWalletChange={onWalletChange} />
  );
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.onWalletChangeHandler();

  expect(onWalletChange).toHaveBeenCalledTimes(1);
  expect(onWalletChange).toHaveBeenCalledWith(true);
});

it('call onWalletChange with false when store is false', () => {
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword walletList={walletList}
      onWalletChange={onWalletChange} />
  );
  wrapper.setState({store: false});
  const instance = wrapper.instance();

  instance.onWalletChangeHandler();

  expect(onWalletChange).toHaveBeenCalledTimes(1);
  expect(onWalletChange).toHaveBeenCalledWith(false);
});

it('it does not call upload when url not set', () => {
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} />);
  wrapper.setState({store: false});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  expect(mockUnlockWallet).toHaveBeenCalledTimes(0);
});

it('shows given walletList as options in dropdown', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({store: false, open: true});

  expect(wrapper.find('option')).toHaveLength(3);
});

it('updates account when option is selected', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('#address').simulate('change', {target:{value: 'asdf'}});

  expect(setState).toHaveBeenCalledWith({address:'asdf'});
});

it('updates the password when typed', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('#password').simulate('change', {target:{value: 'asdf'}});

  expect(setState).toHaveBeenCalledWith({password:'asdf'});
});

it('updates store when checkbox is checked', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('#store').simulate('change', {checked: true});

  expect(setState).toHaveBeenCalledWith({store:true});
});

it('updates store when checkbox is unchecked', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('#store').simulate('change', {checked: false});

  expect(setState).toHaveBeenCalledWith({store:false});
});

it('shows error message when error is true', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true, error: true});

  expect(wrapper.find('.ModalError')).toHaveLength(1);
  expect(wrapper.find('.ModalError').text()).toEqual('Unabled to login. Check your password.');
});

it('does not show error message when error is false', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true, error: false});

  expect(wrapper.find('.ModalError').text()).toHaveLength(0);
});

it('closes the modal when Cancel is pressed', () => {
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword walletList={walletList}/>);
  wrapper.setState({open: true, error: false});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('.flat-cancel').simulate('click');

  expect(setState).toHaveBeenCalledWith({open: false});
});

it('starts unlocking when Unlock is pressed with wallets', () => {
  const url = 'https://localhost:8080';
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword url={url} walletList={walletList}/>);
  wrapper.setState({open: true, error: false});
  const setState = jest.spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();

  wrapper.find('.flat').simulate('click');

  expect(setState).toHaveBeenCalledWith({unlocking: true, error: false});
});

it('unlocks with the values entered in the state for unlockWallet', () => {
  const onUnlock = jest.spyOn(ModalPassword.prototype, 'onUnlock');
  const url = 'https://localhost:8080';
  const walletList = ['asdf','demo','omed'];
  const wrapper = mount(<ModalPassword url={url} walletList={walletList}/>);
  wrapper.setState({open: true, error: false, password: 'password', address:'address'});

  onUnlock.mockClear();

  wrapper.find('.flat').simulate('click');

  expect(onUnlock).toHaveBeenCalledWith('address', 'password');
});

it('creates with the values entered in the state for unlockWallet', () => {
  const createWallet = jest.spyOn(ModalPassword.prototype, 'createWallet');
  const url = 'https://localhost:8080';
  const walletList = [];
  const wrapper = mount(<ModalPassword url={url} walletList={walletList}/>);
  wrapper.setState({open: true, error: false, password: 'password', address:'address'});
  createWallet.mockClear();

  wrapper.find('.flat').simulate('click');

  expect(createWallet).toHaveBeenCalledWith('password');
});

it('does call onWalletChange with false when store is false after unlocking', (done) => {
  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: false});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try {
      expect(onWalletChange).toHaveBeenCalledTimes(1);
      expect(mockUnlockWallet).toHaveBeenCalledTimes(1);
      done();
    } catch(error) {
      done.fail(error);
    }
  },0);
});

it('does call onWalletChange when store is true after unlocking', (done) => {
  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try {
      expect(mockUnlockWallet).toHaveBeenCalledTimes(1);
      expect(onWalletChange).toHaveBeenCalledTimes(1);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('closes the modal when unlock succeeds', (done) => {
  const url = 'https://localhost:8080';
  const close = spyOn(ModalPassword.prototype, 'close');
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try {
      expect(mockUnlockWallet).toHaveBeenCalledTimes(1);
      expect(close).toHaveBeenCalledTimes(1);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('does not close when unlock fails', (done) => {
  const mockBadUnlock = jest.fn().mockImplementation(() => {
    return new Promise((resolve) => {
      resolve(false);
    });
  });
  HttpAccount.mockImplementation(() => {
    return {
      unlockWallet: mockBadUnlock,
    };
  });
  const url = 'https://localhost:8080';
  const close = spyOn(ModalPassword.prototype, 'close');
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try {
      expect(mockBadUnlock).toHaveBeenCalledTimes(1);
      expect(close).toHaveBeenCalledTimes(0);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('sets unlocking:true, error: false when unlock starts', () => {
  const url = 'https://localhost:8080';
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const setState = spyOn(ModalPassword.prototype, 'setState');
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  expect(setState).toHaveBeenCalledWith({unlocking: true, error: false});
});

it('sets unlocking:false, error:false when unlock succeeds', (done) => {
  const url = 'https://localhost:8080';
  const walletList = ['asdf'];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const setState = spyOn(ModalPassword.prototype, 'setState');
  setState.mockClear();
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try{
      expect(setState).toHaveBeenCalledTimes(3);
      expect(setState.calls.argsFor(1)[0]).toEqual({unlocking: false, error: false});
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('does not call onWalletChange when unlock fails', (done) => {
  const mockBadUnlock = jest.fn().mockImplementation(() => {
    return new Promise((resolve) => {
      resolve(false);
    });
  });
  HttpAccount.mockImplementation(() => {
    return {
      unlockWallet: mockBadUnlock,
    };
  });

  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.onUnlock('address', 'password');

  setTimeout(() => {
    try {
      expect(mockBadUnlock).toHaveBeenCalledTimes(1);
      expect(onWalletChange).toHaveBeenCalledTimes(0);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('does call onWalletChange with false when store is false after creating', (done) => {
  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: false});
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try {
      expect(onWalletChange).toHaveBeenCalledTimes(1);
      expect(mockCreateWallet).toHaveBeenCalledTimes(1);
      done();
    } catch (error) {
      done.fail();
    }
  });
});

it('does call onWalletChange when store is true after create', (done) => {
  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try {
      expect(mockCreateWallet).toHaveBeenCalledTimes(1);
      expect(onWalletChange).toHaveBeenCalledTimes(1);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('closes the modal when create succeeds', (done) => {
  const url = 'https://localhost:8080';
  const close = spyOn(ModalPassword.prototype, 'close');
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try {
      expect(mockCreateWallet).toHaveBeenCalledTimes(1);
      expect(close).toHaveBeenCalledTimes(1);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('does not close when create fails', (done) => {
  const mockBadCreate = jest.fn().mockImplementation(() => {
    return new Promise((resolve) => {
      resolve(false);
    });
  });
  HttpAccount.mockImplementation(() => {
    return {
      createWallet: mockBadCreate,
    };
  });
  const url = 'https://localhost:8080';
  const close = spyOn(ModalPassword.prototype, 'close');
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try {
      expect(mockBadCreate).toHaveBeenCalledTimes(1);
      expect(close).toHaveBeenCalledTimes(0);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('sets unlocking:true, error: false when create starts', () => {
  const url = 'https://localhost:8080';
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const setState = spyOn(ModalPassword.prototype, 'setState');
  const instance = wrapper.instance();

  instance.createWallet('password');

  expect(setState).toHaveBeenCalledWith({unlocking: true, error: false});
});

it('sets unlocking:false, error:false when create succeeds', (done) => {
  const url = 'https://localhost:8080';
  const walletList = [];
  const wrapper = mount(<ModalPassword walletList={walletList} url={url}/>);
  wrapper.setState({store: true});
  const setState = spyOn(ModalPassword.prototype, 'setState');
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try{
      expect(setState).toHaveBeenCalledTimes(3);
      expect(setState.calls.argsFor(1)[0]).toEqual({unlocking: false, error: false});
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});

it('does not call onWalletChange when create fails', (done) => {
  const mockBadCreate = jest.fn().mockImplementation(() => {
    return new Promise((resolve) => {
      resolve(false);
    });
  });
  HttpAccount.mockImplementation(() => {
    return {
      createWallet: mockBadCreate,
    };
  });

  const url = 'https://localhost:8080';
  const onWalletChange = jest.fn();
  const walletList = [];
  const wrapper = mount(
    <ModalPassword onWalletChange={onWalletChange}
      walletList={walletList}
      url={url}/>
  );
  wrapper.setState({store: true});
  const instance = wrapper.instance();

  instance.createWallet('password');

  setTimeout(() => {
    try {
      expect(mockBadCreate).toHaveBeenCalledTimes(1);
      expect(onWalletChange).toHaveBeenCalledTimes(0);
      done();
    } catch (error) {
      done.fail(error);
    }
  }, 0);
});
