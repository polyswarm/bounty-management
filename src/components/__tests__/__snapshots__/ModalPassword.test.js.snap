// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`opens with some wallets without crashing 1`] = `
<ModalPassword
  walletList={
    Array [
      "asdf",
      "demo",
    ]
  }
>
  <div
    className="ModalPassword"
  >
    <div
      className="ModalBackground"
      onClick={[Function]}
    />
    <div
      className="ModalContent"
    >
      <header
        className="ModalContentHeader"
      >
        Unlock your wallet.
      </header>
      <form>
        <label
          htmlFor="address"
        >
          Wallet Address
        </label>
        <select
          id="address"
          onChange={[Function]}
          value=""
        >
          <option
            key="asdf"
            value="asdf"
          >
            asdf
          </option>
          <option
            key="demo"
            value="demo"
          >
            demo
          </option>
        </select>
        <label
          htmlFor="password"
        >
          Wallet Password
        </label>
        <input
          id="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <span>
          <input
            id="store"
            onChange={[Function]}
            type="checkbox"
            value={false}
          />
          <label
            htmlFor="store"
          >
            Keep this wallet unlocked
          </label>
        </span>
        <div
          className="ModalError"
        />
      </form>
      <span
        className="Modal-Button-Bar"
      >
        <Button
          disabled={false}
          flat={true}
          onClick={[Function]}
        >
          <button
            className="Button flat"
            disabled={false}
            onClick={[Function]}
          >
            Unlock
          </button>
        </Button>
        <Button
          cancel={true}
          disabled={false}
          flat={true}
          onClick={[Function]}
        >
          <button
            className="Button flat-cancel"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </button>
        </Button>
      </span>
    </div>
  </div>
</ModalPassword>
`;

exports[`opens without crashing 1`] = `
<ModalPassword
  walletList={Array []}
>
  <div
    className="ModalPassword"
  >
    <div
      className="ModalBackground"
      onClick={[Function]}
    />
    <div
      className="ModalContent"
    >
      <header
        className="ModalContentHeader"
      >
        Create a wallet
      </header>
      <form>
        <label
          htmlFor="password"
        >
          Wallet Password
        </label>
        <input
          id="password"
          onChange={[Function]}
          type="password"
          value=""
        />
        <span>
          <input
            id="store"
            onChange={[Function]}
            type="checkbox"
            value={false}
          />
          <label
            htmlFor="store"
          >
            Keep this wallet unlocked
          </label>
        </span>
        <div
          className="ModalError"
        />
      </form>
      <span
        className="Modal-Button-Bar"
      >
        <Button
          disabled={false}
          flat={true}
          onClick={[Function]}
        >
          <button
            className="Button flat"
            disabled={false}
            onClick={[Function]}
          >
            Create
          </button>
        </Button>
        <Button
          cancel={true}
          disabled={false}
          flat={true}
          onClick={[Function]}
        >
          <button
            className="Button flat-cancel"
            disabled={false}
            onClick={[Function]}
          >
            Cancel
          </button>
        </Button>
      </span>
    </div>
  </div>
</ModalPassword>
`;

exports[`renders without crashing 1`] = `
<div
  class="ModalPassword"
>
  
</div>
`;
